DataFormats/Detectors/TRD/src/RawData.cxx

```diff
@@ -84,253 +103,265 @@ uint32_t getlinkerrorflag(const HalfCRUHeader& cruhead, const uint32_t link)
+uint32_t getHCIDFromTrackletHCHeader(const uint32_t& headerword)
 {
-  // make a pretty output of the tracklet.
-  stream << "TrackletMCMData: Raw:0x" << std::hex << tracklet.word << " pos=" << tracklet.pos
-         << "::slope=" << tracklet.slope << "::pid=" << tracklet.pid << "::checkbit="
-         << tracklet.checkbit << std::endl;
-  return stream;
+  TrackletHCHeader header;
+  header.word = headerword;
+  return header.layer * 2 + header.stack * constants::NLAYER * 2 + header.supermodule * constants::NLAYER * constants::NSTACK * 2 + header.side;
 }
-void printTrackletMCMData(o2::trd::TrackletMCMData& tracklet)
+
+uint32_t getChargeFromRawHeaders(const o2::trd::TrackletHCHeader& hcheader, const o2::trd::TrackletMCMHeader* header, const std::array<o2::trd::TrackletMCMData, 3>& data, int pidindex, int trackletindex)
 {
-  LOGF(info, "TrackletMCMData: Raw:0x%08x pos:%d slope:%d pid:0x%08x checkbit:0x%02x",
-       tracklet.word, tracklet.pos, tracklet.slope, tracklet.pid, tracklet.checkbit);
+  uint32_t pid = 0;
+  uint32_t highPID, lowPID; // highPID holds the 8 bits from the mcmheader, and lowPID holds the 12 bits from mcmdata
+  uint32_t datatype = (hcheader.format) >> 2;
+  switch (datatype) {
+    case 0: //Cosmic
+      LOG(error) << "This is a problem cosmic format tracklets ";
+      break;
+    case 1: //TPT
+      LOG(error) << "This is a problem  TPT format tracklets ";
+      break;
+    case 2: //DIS
+      LOG(error) << "This is a problem  DIS format tracklets ";
+      break;
+    case 3:
+      //PID VERSION 1
+      //PID is 20 bits, 8 bits in mcmheader and 12 bits in mcmdata word
+      //frist part of pid (highPID) is in the TrackletMCMHeader
+      //highPID is 7 bits Q2, 1 bit Q1 OR ... 2 bit offset, 6 bits Q2.
+      switch (trackletindex) {
+        case 0:
+          highPID = header->pid0;
+          break;
+        case 1:
+          highPID = header->pid1;
+          break;
+        case 2:
+          highPID = header->pid2;
+          break;
+        default:
+          LOG(warn) << " unknown trackletindex of " << trackletindex << " to " << __func__ << " : " << pidindex;
+          break;
+      }
+      lowPID = data[trackletindex].pid;
+      //lowPID is 6 bits Q0 and 6 bits of Q1
+      int pidword = (highPID << 12) + lowPID;     // the entire original 20 bit pid in the trap chips
+      int dynamicq = hcheader.format & 0x1;       // last bit of format (lsb) defines the version of tracklet charge calculation
+      uint32_t pidoffset = (pidword >> 18) & 0x3; // used for dynamic ranged charge windows.
+      //pidword is here to make this code more readible and less error prone.
+      switch (pidindex) {
+        case 2: //Q2
+          if (!dynamicq) {
+            pid = (pidword >> 13) & 0x7f; // 7 bits at the top of all of pid (MSB)
+          } else {
+            pid = (pidword >> 12) & 0x3f; // 6 bits of Q2 and a shift
+            pid += pidoffset;
```

Comment:
I still don't quite get this scale factor... This would add at most 3 to any pid

Reply:
other end, hence the bitshift of pidoffset a few lines above the quote of your comment.

Reply:
Ah missed that, thanks for clarifying.